### PR TITLE
feat(router): apply filters on payments

### PR DESCRIPTION
**Pull request by apoorvdixit88**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Added functionality to apply list of available filters on payments. One can get the list of filters from _payments/filter_ api and then can apply the required filters (on _payments/list_).

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -1658,30 +1658,43 @@ pub struct PaymentListResponse {
     pub data: Vec<PaymentsResponse>,
 }
 
-#[derive(Clone, Debug, serde::Serialize, ToSchema)]
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct PaymentListFilterConstraints {
+    // The identifier for payment
+    pub payment_id: Option<String>,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:1665]: You can still have these as documentation comments, would help during development.

File: crates/router/src/types/storage/payment_intent.rs [hunk 2]
```diff
@@ -121,4 +133,57 @@ impl PaymentIntentDbExt for PaymentIntent {
...
+        let mut filter = Self::table()
+            .inner_join(payment_attempt::table.on(dsl1::attempt_id.eq(dsl::active_attempt_id)))
+            .filter(dsl::merchant_id.eq(merchant_id.to_owned()))
+            .order(dsl::created_at.desc())
+            .into_boxed();
+
+        match &constraints.payment_id {
+            Some(pid) => {
+                filter = filter.filter(dsl::payment_id.eq(pid.to_owned()));
+            }
```
Comment (SanchithHegde) [inline crates/router/src/types/storage/payment_intent.rs:153]:
```suggestion
            Some(payment_id) => {
                filter = filter.filter(dsl::payment_id.eq(payment_id.to_owned()));
            }
```

Prefer using verbose names.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/router/src/types/storage/payment_intent.rs [hunk 3]
```diff
@@ -121,4 +133,57 @@ impl PaymentIntentDbExt for PaymentIntent {
...
+
+    #[instrument(skip(conn))]
+    async fn apply_filters_on_payments(
+        conn: &PgPooledConn,
+        merchant_id: &str,
+        constraints: &api::PaymentListFilterConstraints,
+    ) -> CustomResult<Vec<(Self, PaymentAttempt)>, errors::DatabaseError> {
+        let offset = constraints.offset.unwrap_or_default();
+        let mut filter = Self::table()
+            .inner_join(payment_attempt::table.on(dsl1::attempt_id.eq(dsl::active_attempt_id)))
```
Comment (ashokkjag) [inline crates/router/src/types/storage/payment_intent.rs:145]: inner join seems the fastest option, but with payment attempt/intent table being the most used table, their size may grow in future and inner join on such large tables may cause a lot of memory pressure on the postgres DB.

Comment (apoorvdixit88) [reply crates/router/src/types/storage/payment_intent.rs:145]: For now I have already used limit, and filtering result for a particular merchant_id!

Some of other ways that I can think of:
indexing: we can make attempt_id/active_attempt column indexed,
limit records and time_range: keeping a lesser limit(10) or limiting it to lesser time_range (not more than 30/60 days)

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/types/storage/payment_intent.rs [hunk 4]
```diff
@@ -179,6 +179,7 @@ impl PaymentIntentDbExt for PaymentIntent {
             filter = filter.filter(dsl1::payment_method.eq_any(payment_method));
         }
 
+        router_env::logger::debug!(filter = %debug_query::<Pg, _>(&filter).to_string());
```
Comment (SanchithHegde) [inline crates/router/src/types/storage/payment_intent.rs:182]: Use `crate::logger` for consistency.

**Review by Narayanbhat166 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**